board/drivers/bxcan.h

```diff
@@ -161,23 +174,35 @@ void can_rx(uint8_t can_number) {
     pending_can_live = 1;
 
     // add to my fifo
-    CAN_FIFOMailBox_TypeDef to_push;
-    to_push.RIR = CAN->sFIFOMailBox[0].RIR;
-    to_push.RDTR = CAN->sFIFOMailBox[0].RDTR;
-    to_push.RDLR = CAN->sFIFOMailBox[0].RDLR;
-    to_push.RDHR = CAN->sFIFOMailBox[0].RDHR;
-
-    // modify RDTR for our API
-    to_push.RDTR = (to_push.RDTR & 0xFFFF000F) | (bus_number << 4);
+    CANPacket_t to_push;
+
+    to_push.returned = 0U;
+    to_push.rejected = 0U;
+    to_push.extended = (CAN->sFIFOMailBox[0].RIR >> 2) & 0x1U;
+    to_push.addr = (to_push.extended != 0) ? (CAN->sFIFOMailBox[0].RIR >> 3) : (CAN->sFIFOMailBox[0].RIR >> 21);
+    to_push.data_len_code = CAN->sFIFOMailBox[0].RDTR & 0xFU;
+    to_push.bus = bus_number;
+    to_push.data[0] = CAN->sFIFOMailBox[0].RDLR & 0xFFU;
+    to_push.data[1] = (CAN->sFIFOMailBox[0].RDLR >> 8) & 0xFFU;
+    to_push.data[2] = (CAN->sFIFOMailBox[0].RDLR >> 16) & 0xFFU;
+    to_push.data[3] = (CAN->sFIFOMailBox[0].RDLR >> 24) & 0xFFU;
+    to_push.data[4] = CAN->sFIFOMailBox[0].RDHR & 0xFFU;
+    to_push.data[5] = (CAN->sFIFOMailBox[0].RDHR >> 8) & 0xFFU;
+    to_push.data[6] = (CAN->sFIFOMailBox[0].RDHR >> 16) & 0xFFU;
+    to_push.data[7] = (CAN->sFIFOMailBox[0].RDHR >> 24) & 0xFFU;
 
     // forwarding (panda only)
     int bus_fwd_num = (can_forwarding[bus_number] != -1) ? can_forwarding[bus_number] : safety_fwd_hook(bus_number, &to_push);
     if (bus_fwd_num != -1) {
-      CAN_FIFOMailBox_TypeDef to_send;
-      to_send.RIR = to_push.RIR | 1; // TXRQ
-      to_send.RDTR = to_push.RDTR;
-      to_send.RDLR = to_push.RDLR;
-      to_send.RDHR = to_push.RDHR;
+      CANPacket_t to_send;
+
+      to_send.returned = 0U;
+      to_push.rejected = 0U;
+      to_send.extended = to_push.extended; // TXRQ
+      to_send.addr = to_push.addr;
+      to_send.bus = to_push.bus;
+      to_send.data_len_code = to_push.data_len_code;
+      (void)memcpy(to_send.data, to_push.data, dlc_to_len[to_push.data_len_code]);
```

Comment:
do we sanity check the `data_len_code` somewhere? what if we send a packet where the DLC is larger than the actual data buffer we sent?